### PR TITLE
Support deleting and renaming files

### LINKED ISSUE #129: Delete files
Offer the option to delete files.

### LINKED ISSUE #130: Rename files
Offer the option to rename files.

### PR DESCRIPTION
**Pull request by MarcelGarus**
**Merge note:** This branch is based on the `fix-caching` branch, so #160 should be merged first.

Fixes #129 
Fixes #130 

By long-tapping on files, a menu appears that lets the user select actions to take. The <kbd>Move</kbd> action does nothing for now, but <kbd>Delete</kbd> and <kbd>Rename</kbd> already work.

`FileMenu` | `RenameDialog`
---|---
image | image

**Note:** Although this technically counts as being able to delete and rename files, this leaves a lot to be desired. For example, after selecting an option, the user needs to navigate away from the screen and then again to the screen to cause a reload and the files to get renamed/disappear. That definitely needs to change before release. This is a more general functionality that needs to be implemented on the level of the `HiveCache` and therefore shouldn't be part of this PR.

**Todo before merging:**
- [x] add title to `RenameDialog`
- [x] internationalize and localize newly added Strings

### CODE REVIEW

File: lib/file/data.dart [hunk 1]
```diff
@@ -106,4 +107,38 @@ class File implements Entity<File>, Comparable<File> {
...
+      'newName': newName,
+    });
+    copyWith(name: newName).saveToCache();
+  }
+
+  Future<void> moveTo(Id<File> parentDirectory) async {
+    await services.api.patch('fileStorage/$id', body: {
+      'parent': parentDirectory,
+    });
+  }
```
Comment (JonasWanke) [inline lib/file/data.dart]: I'd guess that these network calls return the modified `File`-object. Can we store that directly?

File: lib/file/data.dart [hunk 2]
```diff
@@ -106,4 +107,38 @@ class File implements Entity<File>, Comparable<File> {
...
+
+  Future<void> delete() => services.api.delete('fileStorage/$id');
```
Comment (JonasWanke) [inline lib/file/data.dart:144]: We should probably delete this file from our cache as well

File: lib/l10n/intl_de.arb [hunk 3]
```diff
@@ -119,5 +115,15 @@
...
+  "file_fileMenu_createdAt": "Erstellt: {createdAt}",
+  "file_fileMenu_modifiedAt": "Zuletzt geändert: {modifiedAt}",
+  "file_fileMenu_makeAvailableOffline": "Offline verfügbar machen",
+  "file_fileMenu_delete": "Löschen",
+  "file_fileMenu_rename": "Umbenennen",
+  "file_fileMenu_open": "Öffnen",
+  "file_fileMenu_move": "Verschieben",
+  "file_renameDialog_title": "Datei umbenennen",
+  "file_renameDialog_inputHint": "Neuer Dateiname",
+  "file_renameDialog_rename": "Umbenennen"
```
Comment (JonasWanke) [inline lib/l10n/intl_de.arb]: Please keep these sorted alphabetically (an issue on L42n was created: JonasWanke/l42n#8)

File: lib/l10n/intl_en.arb [hunk 4]
```diff
@@ -119,5 +115,15 @@
...
+  "file_fileMenu_createdAt": "Created: {createdAt}",
+  "file_fileMenu_modifiedAt": "Last modified: {modifiedAt}",
+  "file_fileMenu_makeAvailableOffline": "Make available offline",
+  "file_fileMenu_delete": "Delete",
+  "file_fileMenu_rename": "Rename",
+  "file_fileMenu_open": "Open",
+  "file_fileMenu_move": "Move",
+  "file_renameDialog_title": "Rename file",
+  "file_renameDialog_inputHint": "New file name",
+  "file_renameDialog_rename": "Rename"
```
Comment (JonasWanke) [inline lib/l10n/intl_en.arb]: Please keep these sorted alphabetically (an issue on L42n was created: JonasWanke/l42n#8)

File: lib/file/widgets/file_menu.dart [hunk 5]
```diff
@@ -0,0 +1,117 @@
+import 'package:black_hole_flutter/black_hole_flutter.dart';
+import 'package:flutter/material.dart';
+import 'package:pedantic/pedantic.dart';
+import 'package:schulcloud/app/app.dart';
+
+import '../data.dart';
+import 'file_thumbnail.dart';
+
+class FileMenu extends StatelessWidget {
+  const FileMenu({this.file});
```
Comment (JonasWanke) [inline lib/file/widgets/file_menu.dart]:
```suggestion
  const FileMenu(this.file) : assert(file != null);
```

File: lib/file/widgets/file_tile.dart [hunk 6]
```diff
@@ -14,38 +15,10 @@ class FileTile extends StatelessWidget {
...
-            title: Text(s.file_fileTile_details_offline),
-            trailing: Switch.adaptive(
-              value: false,
-              onChanged: (_) {},
-            ),
-          ),
-        ],
-      ),
+      builder: (context) => FileMenu(file: file),
     );
```
Comment (JonasWanke) [inline lib/file/widgets/file_tile.dart]: - This should probably be a static function on `FileMenu` for better reuse.
- WIth `black_hole_flutter` you can use `context.showFancyModalBottomSheet`

File: lib/file/widgets/file_menu.dart [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  @override
+  void initState() {
+    _controller = TextEditingController(text: widget.oldName);
+    super.initState();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return AlertDialog(
+      title: Text('Rename file'),
```
Comment (JonasWanke) [inline lib/file/widgets/file_menu.dart]: L10n

File: lib/l10n/intl_en.arb [hunk 8]
```diff
@@ -119,5 +115,15 @@
...
-  "signIn_signInScreen_placeholder": "There could go some other information down here."
+  "signIn_signInScreen_placeholder": "There could go some other information down here.",
+  "file_fileMenu_createdAt": "Created: {createdAt}",
+  "file_fileMenu_modifiedAt": "Last modified: {modifiedAt}",
+  "file_fileMenu_makeAvailableOffline": "Make available offline",
+  "file_fileMenu_delete": "Delete",
+  "file_fileMenu_rename": "Rename",
+  "file_fileMenu_open": "Open",
+  "file_fileMenu_move": "Move",
+  "file_renameDialog_title": "Rename file",
```
Comment (JonasWanke) [inline lib/l10n/intl_en.arb]: For titles, you can omit the suffix `_title` and just use `file_renameDialog`

File: lib/file/widgets/file_menu.dart [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+      builder: (_) => RenameDialog(oldName: file.name),
+    );
+
+    if (newName == null) {
+      return;
+    }
+
+    unawaited(file.rename(newName));
+    context.navigator.pop();
+  }
```
Comment (JonasWanke) [inline lib/file/widgets/file_menu.dart]: One method uses `context.navigator` and the other one uses `context.rootNavigator`? AFAIK both refer to the same `Navigator` in this case, so probably prefer `context.navigator`?
```suggestion
  Future<void> _delete(BuildContext context) async {
    unawaited(file.delete());
    context.navigator.pop();
  }

  Future<void> _rename(BuildContext context) async {
    final newName = await showDialog<String>(
      context: context,
      useRootNavigator: true,
      builder: (_) => RenameDialog(oldName: file.name),
    );

    if (newName == null) {
      return;
    }

    unawaited(file.rename(newName));
    context.navigator.pop();
  }
```

**Review by JonasWanke — CHANGES_REQUESTED** [review]
While it works, we should improve the UI:
- Add icons for the different options in `FileMenu` and maybe order them differently (e.g. rename as a pen-icon next to the file's name)
- Add a confirm dialog before deleting a file
- Use `SecondaryButton` and `PrimaryButton` for cancel and rename actions in the dialog
- Add snack bars after both rename and delete as a confirmation for the user and maybe some loading indicator while the network request is in progress.
- Nice to have: I guess that most users intend to only change the name and not the file extension when selecting "Rename". Maybe place the cursor before the last dot or even pre-select the name (without extension)?

FYI: I created #214 in response to your note

File: lib/file/data.dart (hunk 2 above)
Comment (MarcelGarus) [reply lib/file/data.dart:144]: HiveCache currently doesn't support removing items… I opened an issue for that: https://github.com/schul-cloud/schulcloud-flutter/issues/216

File: lib/app/services/snack_bar.dart [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+  SnackBarRequest({@required this.completer, @required this.snackBar})
+      : assert(completer != null),
+        assert(snackBar != null);
+
+  final Completer<ScaffoldFeatureController<SnackBar, SnackBarClosedReason>>
+      completer;
+  final SnackBar snackBar;
+}
+
+/// A service that offers displaying app-wide SnackBars at the bottom.
```
Comment (JonasWanke) [inline lib/app/services/snack_bar.dart]:
```suggestion
/// A service that offers displaying app-wide [SnackBar]s at the bottom.
```

File: lib/app/services/snack_bar.dart [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    return completer.future;
+  }
+
+  Future<ScaffoldFeatureController<SnackBar, SnackBarClosedReason>> showMessage(
+    String message,
+  ) =>
+      show(SnackBar(content: Text(message)));
+
+  Future<ScaffoldFeatureController<SnackBar, SnackBarClosedReason>>
+      showMessageShortly(String message) {
```
Comment (JonasWanke) [inline lib/app/services/snack_bar.dart]: "Shortly" means soon, but I guess you mean "briefly"

File: lib/app/widgets/form.dart [hunk 12]
```diff
@@ -32,7 +32,7 @@ extension FormDialogs on BuildContext {
   Future<bool> showConfirmDeleteDialog(String message) {
     final s = this.s;
 
-    return showDialog<bool>(
+    final result = showDialog<bool>(
```
Comment (JonasWanke) [inline lib/app/widgets/form.dart:36]: Can you please implement this fix for `showDiscardChangesDialog` (defined above) as well?

File: lib/file/widgets/file_menu.dart [hunk 13]
```diff
@@ -27,8 +31,32 @@ class FileMenu extends StatelessWidget {
       return;
     }
 
-    unawaited(file.rename(newName));
     context.navigator.pop();
+    await services.snackBar.performAction(
+      action: () => file.rename(newName),
+      loadingMessage: 'Renaming ${file.name} to $newName…',
+      successMessage: 'Renamed to $newName 😊',
+      failureMessage: "Couldn't rename to $newName",
```
Comment (JonasWanke) [inline lib/file/widgets/file_menu.dart]: I18n

File: lib/file/widgets/file_menu.dart [hunk 14]
```diff
@@ -27,8 +31,32 @@ class FileMenu extends StatelessWidget {
...
+    if (!confirmed) {
+      return;
+    }
+
+    context.navigator.pop();
+    await services.snackBar.performAction(
+      action: file.delete,
+      loadingMessage: 'Deleting ${file.name}…',
+      successMessage: 'Deleted ${file.name} 😊',
+      failureMessage: "Couldn't delete ${file.name}",
```
Comment (JonasWanke) [inline lib/file/widgets/file_menu.dart]: I18n

**Review by JonasWanke — CHANGES_REQUESTED** [review]

**Review by JonasWanke — APPROVED** [review]

**Pull request merged**